### PR TITLE
Fix adding disks to more than one scsi adapter. Fixes #42399

### LINKED ISSUE #42399: vSphere: Cannot add disks to more than one scsi adapter (disk count per node > 16)
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`): v1.4.x v1.5.x v1.6.x master

**Environment**: vSphere based install with vSphere Cloud Provider enabled.
- **Cloud provider or hardware configuration**: vSphere
- **OS** (e.g. from /etc/os-release): NA
- **Kernel** (e.g. `uname -a`):NA
- **Install tools**:kubernetes-anywhere
- **Others**:

**What happened**: vSphere Cloud Provider configures additional SCSI adapters to scale the number of volumes that can be attached to a VM (1 adapter = 16 disks). The code within the cloud provider does not handle this correctly. Additional disks attach successfully to the VM but when k8s verifies the attach status it fails leading to k8s retrying the attach that never succeeds. In addition to this when the SCSI adapter is created for the first time the code does not use it to attach the disk leading to a failed attach disk request to the cloud provider.

**What you expected to happen**: Additional disks should be added to the VM upto the max number supported. 

**How to reproduce it** (as minimally and precisely as possible): Create a single node cluster and create a statefulset with > 16 replicas or have a pod with more than 16 disks.

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by kerneltime**
**What this PR does / why we need it**: Allows a single node to use more than 16 disks.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #42399

**Special notes for your reviewer**: 

**Release note**:

```release-note
Fix adding disks to more than one scsi adapter.
```

### CODE REVIEW

**Comment by kerneltime** [comment]
This needs to be cherry picked to 1.6 1.5 and 1.4

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 1]
```diff
@@ -761,7 +761,15 @@ func (vs *VSphere) AttachDisk(vmDiskPath string, nodeName k8stypes.NodeName) (di
 			return "", "", err
 		}
 
-		scsiController = getSCSIController(vmDevices, vs.cfg.Disk.SCSIControllerType)
+		// Get VM device list
+		_, vmDevices, _, err := getVirtualMachineDevices(ctx, vs.cfg, vs.client, vSphereInstance)
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: L:757-762 ->  not needed.

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 2]
```diff
@@ -761,7 +761,15 @@ func (vs *VSphere) AttachDisk(vmDiskPath string, nodeName k8stypes.NodeName) (di
...
+		if err != nil {
+			glog.Errorf("cannot get vmDevices for VM")
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Can we print or return err?
glog.Errorf("cannot get vmDevices for VM. err: %s", err)

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 1 above)
Comment (kerneltime) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: It is needed vmDevices get written over in line 751 and needs to be reset to all devices.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 2 above)
Comment (kerneltime) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: will do

**Comment by kerneltime** [comment]
@k8s-bot gci gke e2e test this
@k8s-bot node e2e test this
@k8s-bot cvm gke e2e test this

**Comment by kerneltime** [comment]
@dchen1107 this is an important fix that needs to go into 1.6 (without this any node selected for a pod with more then 16 disks will result in pod never powering on apps such as those in stateful sets to be hung)

**Comment by divyenpatel** [comment]
lgtm

**Comment by kerneltime** [comment]
@k8s-bot gci gke e2e test this
@k8s-bot cvm gce e2e test this

**Comment by kerneltime** [comment]
@k8s-bot gci gke e2e test this
@k8s-bot cvm gce e2e test this

**Comment by divyenpatel** [comment]
Tested this change with mounting 30 volumes on the pod. Verified 3 SCSI adapters found on node VM.

```
[ root@pod30volumes:/ ]$ mount | grep "thinclaim"                              
/dev/sdp on /thinclaim2 type ext4 (rw,relatime,data=ordered)
/dev/sdb on /thinclaim5 type ext4 (rw,relatime,data=ordered)
/dev/sdw on /thinclaim6 type ext4 (rw,relatime,data=ordered)
/dev/sdad on /thinclaim25 type ext4 (rw,relatime,data=ordered)
/dev/sdo on /thinclaim29 type ext4 (rw,relatime,data=ordered)
/dev/sdh on /thinclaim19 type ext4 (rw,relatime,data=ordered)
/dev/sdm on /thinclaim20 type ext4 (rw,relatime,data=ordered)
/dev/sds on /thinclaim9 type ext4 (rw,relatime,data=ordered)
/dev/sdq on /thinclaim11 type ext4 (rw,relatime,data=ordered)
/dev/sdl on /thinclaim14 type ext4 (rw,relatime,data=ordered)
/dev/sdr on /thinclaim27 type ext4 (rw,relatime,data=ordered)
/dev/sdi on /thinclaim28 type ext4 (rw,relatime,data=ordered)
/dev/sdac on /thinclaim1 type ext4 (rw,relatime,data=ordered)
/dev/sdae on /thinclaim7 type ext4 (rw,relatime,data=ordered)
/dev/sdn on /thinclaim16 type ext4 (rw,relatime,data=ordered)
/dev/sdab on /thinclaim26 type ext4 (rw,relatime,data=ordered)
/dev/sdt on /thinclaim13 type ext4 (rw,relatime,data=ordered)
/dev/sdv on /thinclaim24 type ext4 (rw,relatime,data=ordered)
/dev/sdf on /thinclaim23 type ext4 (rw,relatime,data=ordered)
/dev/sdc on /thinclaim30 type ext4 (rw,relatime,data=ordered)
/dev/sdu on /thinclaim8 type ext4 (rw,relatime,data=ordered)
/dev/sdj on /thinclaim10 type ext4 (rw,relatime,data=ordered)
/dev/sdd on /thinclaim15 type ext4 (rw,relatime,data=ordered)
/dev/sdg on /thinclaim18 type ext4 (rw,relatime,data=ordered)
/dev/sde on /thinclaim21 type ext4 (rw,relatime,data=ordered)
/dev/sdz on /thinclaim3 type ext4 (rw,relatime,data=ordered)
/dev/sdaa on /thinclaim4 type ext4 (rw,relatime,data=ordered)
/dev/sdx on /thinclaim17 type ext4 (rw,relatime,data=ordered)
/dev/sdy on /thinclaim22 type ext4 (rw,relatime,data=ordered)
/dev/sdk on /thinclaim12 type ext4 (rw,relatime,data=ordered)
[ root@pod30volumes:/ ]$ mount | grep "thinclaim" | wc -l                      
30
[ root@pod30volumes:/ ]$
```

**Comment by kerneltime** [comment]
CC @dchen1107

**Pull request merged**